modules/onewire/onewire.c

```diff
@@ -0,0 +1,418 @@
+/**
+ * |----------------------------------------------------------------------
+ * | Copyright (C) Tilen Majerle, 2014
+ * |
+ * | This program is free software: you can redistribute it and/or modify
+ * | it under the terms of the GNU General Public License as published by
+ * | the Free Software Foundation, either version 3 of the License, or
+ * | any later version.
+ * |
+ * | This program is distributed in the hope that it will be useful,
+ * | but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * | GNU General Public License for more details.
+ * |
+ * | You should have received a copy of the GNU General Public License
+ * | along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * |----------------------------------------------------------------------
+ */
+
+#include "onewire.h"
+
+// ONEWIRE ABSTRACTION HELPERS
+
+// NOTE(vwnguyen): Blocking sleep function. Seems to be OK for our application.
+// NOTE(vwnguyen): Seems to be more accurate than using ChThdSleepMicroseconds()
+void ONEWIRE_DELAY(uint32_t time_us) {
+    usleep(time_us);
+}
+
+uint8_t ONEWIRE_LOW(OneWire_t *OneWireStruct ){
+    if (!OneWireStruct) return 0;
+    palClearLine(OneWireStruct->PalLine);
+    return 1;
+}
+
+uint8_t ONEWIRE_HIGH(OneWire_t *OneWireStruct) {
+    if (!OneWireStruct) return 0;
+    palSetLine(OneWireStruct->PalLine); 
+    return 1;
+}
+
+// NOTE(vwnguyen): Min time taken to toggle between OUTPUT -> INPUT ~= 4-5us
+// NOTE(vwnguyen): using ONEWIRE_DELAY on top of this toggle time should result
+// NOTE(vwnguyen): in total_delay = toggle_time + delay_time
+uint8_t ONEWIRE_INPUT(OneWire_t *OneWireStruct) {
+    if (!OneWireStruct) return 0;
+    palSetLineMode(OneWireStruct->PalLine, PAL_MODE_INPUT | PAL_STM32_OSPEED_HIGHEST);
+    return 1;
+}
+
+uint8_t ONEWIRE_OUTPUT(OneWire_t *OneWireStruct) {
+    if (!OneWireStruct) return 0;
+    palSetLineMode(OneWireStruct->PalLine, PAL_STM32_MODE_OUTPUT | PAL_STM32_OSPEED_HIGHEST);
+    return 1;
+}
+
+// ONEWIRE IMPLEMENTATION
+
+uint8_t OneWire_Init(OneWire_t* OneWireStruct, uint32_t PalLine) {
+    if (!OneWireStruct) return 0;
+    OneWireStruct->PalLine = PalLine;
+    return 1;
+}
+
+char OneWire_Reset(OneWire_t* OneWireStruct) {
+    if (!OneWireStruct) return -1;
+    uint8_t i;
+
+    /* Line low, and wait */
+    ONEWIRE_OUTPUT(OneWireStruct);
+    ONEWIRE_LOW(OneWireStruct);
+    ONEWIRE_DELAY(ONEWIRE_TX_MIN_RESET_PULSE_TIME_USEC);
+    
+    /* Release line and wait */
+    ONEWIRE_INPUT(OneWireStruct);
+    ONEWIRE_DELAY(ONEWIRE_WAIT_PRESENCE_PULSE_TIME_USEC);
+    
+    /* Check bit value */
+    i = palReadLine(OneWireStruct->PalLine);
+
+    /* Complete reset pulse min time */
+    ONEWIRE_DELAY(ONEWIRE_RX_MIN_RESET_PULSE_TIME_USEC);
+    /* Return value of presence pulse, 0 = OK, 1 = ERROR */
+    return i;

```

Comment:
char allows for returning -1 as a specific usage exit code, but at this point its too much work to maintain. So I am inclined to change it to 1 for success 0 for failure in all cases (usage or presence pulse low in this case)

Reply:
This may work for this example, however for something a function like ReadByte, 

it may actually read bus a value of 0 or 1, so -1 makes more sense as a return value to distinguish a specific error. Maybe the best solution is to return a max uint8_t value of 255 on failure, and accept that the value may be correct or an error.

Reply:
I suggest you use a enum type (or #define) as a return from this functions (and the others like it)

```
typedef enum {
    ONEWIRE_IO_STATUS_ERROR = -1,
    ONEWIRE_IO_STATUS_LOW   =  0,
    ONEWIRE_IO_STATUS_HI    =  1,
} OneWireIoStatus;
```